### PR TITLE
[FLINK-27015][hive] Fix exception for casting timestamp to decimal in Hive dialect

### PR DESCRIPTION
**Pull request by luoyuxia**
## What is the purpose of the change
To fix the exception for casting timestamp to decimal in Hive dialect

## Brief change log
  - Add a custome udf `flink_hive_to_decimal ` to convert other datatypes include timestamp to decimal
  - When it's for insert timestamp to decimal, make it to be a `RexCall` to call function `flink_hive_to_decimal`

## Verifying this change
Added test in `HiveDialectQueryITCase#testCastTimeStampToDecimal`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / **not documented**)

### CODE REVIEW

**Comment by wuchong** [comment]
The `HiveDialectQueryITCase.testCastTimeStampToDecimal` is failed.

File: flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/connectors/hive/HiveDialectQueryITCase.java [hunk 1]
```diff
@@ -743,6 +743,31 @@ public void testBoolComparison() {
...
+                    CollectionUtil.iteratorToList(
+                            tableEnv.executeSql(
+                                            "select cast(cast('2012-12-19 11:12:19.1234567' as timestamp) as decimal(30,8))")
+                                    .collect());
+            assertThat(results.toString()).isEqualTo("[+I[1355886739.12345670]]");
+
+            // test insert timestamp type to decimal type directly
+            tableEnv.executeSql("create table t1 (c1 DECIMAL(38,6))");
+            tableEnv.executeSql("create table t2 (c2 TIMESTAMP)");
+            tableEnv.executeSql("insert into t2 values('2029-06-05 14:28:29.23445')").await();
```
Comment (wuchong) [inline flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/connectors/hive/HiveDialectQueryITCase.java:760]: Shouldn't it be inserting values into `t1` for testing `insert timestamp type to decimal type directly`?

Comment (luoyuxia) [reply flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/connectors/hive/HiveDialectQueryITCase.java]: Yeah, that's what's in my mind. But miss it by mistake.

**Review by wuchong — APPROVED** [review]
LGTM.

**Pull request merged**